### PR TITLE
FIX: make safe to add / remove artists during ArtistList iteration

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

If you remove artists during iteration some of the artists will be skipped due
to the underlying list being mutated.

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

### CODE REVIEW

**Review by QuLogic — APPROVED** [review]

**Review by anntzer — CHANGES_REQUESTED** [review]
But this seems to be normal Python semantics, no? ("if you want to iterate on a list and remove entries from it at the same time, you need to make a copy first")  I'm not sure we should deviate from standard semantics here (especially as the name ArtistList does suggest list-like semantics.

Hence blocking as I think this at least needs some more discussion, but open to changing my mind.

**Review by timhoffm — APPROVED** [review]
`ArtistList` exists to make the returned list finally immutable.

If we'll do this using a tuple, as suggested in the `ArtistList` docstring, that will effectively be equivalent to the behavioral change in this PR. So anticipating the change already is fine.

One could imagine alternative scenarios with a read-only proxy object to the internal data, which would suffer the modification problem. Such an object would only make sense if creating a copy is too expensive, but I don't think that's the case.

**Comment by tacaswell** [comment]
> But this seems to be normal Python semantics, no?

I had the same thought, however because it is a proxy which is providing a list-like interface of a filtered view on top of another list I think we should err on the side of hiding the mutation.  Because all of the proxy lists operate on top of the same underlying list you could come up with a pathological example where you are iterating over and mutating more than one and they start to affect each other.

**Comment by anntzer** [comment]
Even immutable proxies on mutable containers don't copy the entries, e.g. `d = {<some dict>}; for k in d: d.pop(k)` fails.  Still, I don't feel that strongly about this, as long as this has been thought out.

**Comment by tacaswell** [comment]
Moved to 3.6 and I'll add a behavior change note?

File: doc/api/next_api_changes/behavior/22229-TAC.rst [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+children combined`), a copy of the contents is made.  This ensure that artists
+can safely be added or removed from the Axes while iterating over their children.
+
+This is a departure from the expected behavior of mutable iterable data types
+in Python -- iterating over list while mutating it has surprising consequences
+and dictionaries will error if they change size during iteration.  In this case
+because all of the accessors are filtered views of the same underlying list, it
+is possible for seemingly unrelated changes, such as removing a Line while
+iterating over patches, we have opted to make a copy of the relevant children
+before yielding them to the user.
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/22229-TAC.rst:15]: It is possible for seemingly unrelated changes to what? There seems to be a clause missing.

Comment (tacaswell) [reply doc/api/next_api_changes/behavior/22229-TAC.rst]: Is that any clearer?

File: doc/api/next_api_changes/behavior/22229-TAC.rst [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+children combined`), a copy of the contents is made.  This ensure that artists
+can safely be added or removed from the Axes while iterating over their children.
+
+This is a departure from the expected behavior of mutable iterable data types
+in Python -- iterating over list while mutating it has surprising consequences
+and dictionaries will error if they change size during iteration.  In this case
+because all of the accessors are filtered views of the same underlying list, it
+is possible for seemingly unrelated changes, such as removing a Line, to affect
+the to affect the iteration over any of the other accessors, we have opted to
+make a copy of the relevant children before yielding them to the user.
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/22229-TAC.rst:15]:
```suggestion
and dictionaries will error if they change size during iteration.
Because all of the accessors are filtered views of the same underlying list, it
is possible for seemingly unrelated changes, such as removing a Line, to affect
the iteration over any of the other accessors. In this case, we have opted to
make a copy of the relevant children before yielding them to the user.
```
